src/comms/config_layer.cpp

```diff
@@ -184,3 +240,173 @@ void Config::fill_data(CommsPacket packets[MAX_CONFIG_PACKETS], uint8_t sizes[MA
+        return false;
+    }
+
+    return true;
+}
+
+bool ConfigLayer::store_config() {
+    // initialize: erase config.pack if exists, reopen
+    // check return values of everything!!!
+    if (sdcard.exists("/config.pack")) {
+        if (sdcard.rm("/config.pack") != 0) {
+            if (!CONFIG_ERR_HANDLER(CONFIG_RM_FAIL)) return false;
+        }
+    }
+    if (sdcard.touch("/config.pack") != 0) {
+        if (!CONFIG_ERR_HANDLER(CONFIG_TOUCH_FAIL)) return false;
+    }
+    if (sdcard.open("/config.pack") != 0) {
+        if (!CONFIG_ERR_HANDLER(CONFIG_OPEN_FAIL)) return false;
+    }
+
+    // total size of /config.pack: MAX_CONFIG_PACKETS * (sizeof(CommsPacket) + 1)
+    // num of packets * size of each packet == num of bytes for all packets
+    int config_byte_size = MAX_CONFIG_PACKETS * sizeof(CommsPacket);
+
+    // calculate checksum on config packet array
+#ifdef CONFIG_LAYER_DEBUG
+    Serial.printf("store_config: computing checksum for received config\n");
+#endif
+    uint8_t* config_bytes = (uint8_t*)config_packets;
+    uint64_t checksum = sd_checksum64(config_bytes, config_byte_size);
+    checksum += sd_checksum64(subsec_sizes, MAX_CONFIG_PACKETS);
+
+    // write robot id byte to ref data so it can be compared directly
+    sdcard.write((uint8_t*)(&config.robot_id), sizeof(float));
+
+    Serial.printf("Robot ID from config: %d\n", (int)config.robot_id);
+
+    for (int i = 0; i < MAX_CONFIG_PACKETS; i++) {
+        // write checksum once
+        sdcard.write((uint8_t*)(&checksum), sizeof(uint64_t)); // reinterpret addr of checksum as byte array
+        // write one packet
+        sdcard.write((uint8_t*)(&config_packets[i]), sizeof(CommsPacket));
+    }
+    sdcard.write(subsec_sizes, MAX_CONFIG_PACKETS);
+
+    sdcard.close();
+
+    Serial.printf("Finished writing to SD\n");
+
+    return true;
+}
+
+uint64_t ConfigLayer::sd_checksum64(uint8_t* arr, uint64_t n) {
+    uint64_t out = 0;
+    for (uint32_t i = 0; i < n; i++) {
+        out += arr[i];
+    }
+#ifdef CONFIG_LAYER_DEBUG
+    Serial.printf("sd_checksum64: returned %x\n", out);
+#endif
+    return out;
+}
+
+bool ConfigLayer::CONFIG_ERR_HANDLER(int err_code) {
+    // vars that may be used by err handler procedures
+    float prev_time, delta_time;
+
+    // make sure that every error case returns a value!
+    switch (err_code) {
+    case CONFIG_RM_FAIL:
+        Serial.println("CONFIG_ERROR::config failed to remove /config.pack from SD");
+        return false;
+
+    case CONFIG_TOUCH_FAIL:
+        Serial.println("CONFIG_ERROR::config failed to create /config.pack on SD");
+        return false;
+
+    case CONFIG_OPEN_FAIL:
+        Serial.println("CONFIG_ERROR::config failed to open /config.pack from SD");
+        return false;
+
+    case CONFIG_ID_MISMATCH:
+        prev_time = (float)millis()/1000;
+        delta_time = (float)millis()/1000 - prev_time;
+        while (1) {
+            if (delta_time >= 2.0f) {
```

Comment:
millis() returns a uint32_t. Just use the integer value. No need to convert to float

Reply:
I converted it to float so that I could measure the time in seconds. However, I realize that I can simply measure in milliseconds to achieve the same thing- so I will go ahead and change that. (Just wanted to explain briefly why I did it so I don't look fully insane)